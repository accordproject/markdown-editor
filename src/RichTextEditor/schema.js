export const LIST_ITEM = 'list_item';
export const BLOCK_QUOTE = 'block_quote';
export const OL_LIST = 'ol_list';
export const UL_LIST = 'ul_list';
export const PARAGRAPH = 'paragraph';
export const H1 = 'heading_one';
export const H2 = 'heading_two';
export const H3 = 'heading_three';
export const H4 = 'heading_four';
export const H5 = 'heading_five';
export const H6 = 'heading_six';
export const MARK_BOLD = 'bold';
export const MARK_ITALIC = 'italic';
export const MARK_CODE = 'code';
export const MARK_HTML = 'html';
export const HR = 'horizontal_rule';
export const LINK = 'link';
export const IMAGE = 'image';
export const HTML_BLOCK = 'html_block';
export const CODE_BLOCK = 'code_block';

export const HTML_INLINE = 'html_inline';
export const SOFTBREAK = 'softbreak';
export const LINEBREAK = 'linebreak';


export const LIST_TYPES = [OL_LIST, UL_LIST];

export const INLINES = [LINEBREAK, SOFTBREAK, HTML_INLINE, LINK, IMAGE];
export const VOIDS = [LINEBREAK, SOFTBREAK, IMAGE, HR];
export const HEADINGS = [H1, H2, H3, H4, H5, H6];

export const MARKS = [MARK_BOLD, MARK_ITALIC, MARK_CODE, MARK_HTML];

const withSchema = (editor) => {
  const { isVoid, isInline } = editor;
  editor.isVoid = element => (VOIDS.includes(element.type) ? true : isVoid(editor));
  editor.isInline = element => (INLINES.includes(element.type) ? true : isInline(editor));

  return editor;
};

export default withSchema;
