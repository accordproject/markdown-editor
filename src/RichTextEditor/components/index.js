import React from 'react';
import PropTypes from 'prop-types';
import ImageElement from '../plugins/withImages';
import {
  Code, Html, ListItem, OLList, Quote, ULList
} from './Block';
import {
  Heading, Inline, Link, Paragraph
} from './Node';
import { HorizontalRule, Linebreak, Softbreak } from './Span';
import {
  PARAGRAPH, LINK, IMAGE, H1, H2, H3, H4, H5, H6, HR,
  CODE_BLOCK, HTML_BLOCK, BLOCK_QUOTE, UL_LIST, OL_LIST, LIST_ITEM,
  HTML_INLINE, SOFTBREAK, LINEBREAK, HEADINGS
} from '../utilities/schema';
import generateId from '../utilities/generateId';

const Element = (props) => {
  const {
    attributes, children, element, customElements
  } = props;
  const { type, data } = element;
  const headingId = HEADINGS.includes(type) ? generateId(element): null;
  const baseElementRenderer = {
    [PARAGRAPH]: () => (<Paragraph {...attributes}>{children}</Paragraph>),
    [H1]: () => (<Heading id={headingId} as="h1" {...attributes}>{children}</Heading>),
    [H2]: () => (<Heading id={headingId} as="h2" {...attributes}>{children}</Heading>),
    [H3]: () => (<Heading id={headingId} as="h3" {...attributes}>{children}</Heading>),
    [H4]: () => (<Heading id={headingId} as="h4" {...attributes}>{children}</Heading>),
    [H5]: () => (<Heading id={headingId} as="h5" {...attributes}>{children}</Heading>),
    [H6]: () => (<Heading id={headingId} as="h6" {...attributes}>{children}</Heading>),
    [SOFTBREAK]: () => (<Softbreak {...attributes}>{children}</Softbreak>),
    [LINEBREAK]: () => (<Linebreak {...attributes} />),
    [LINK]: () => (<Link {...attributes} href={data.href}>{children}</Link>),
    [HTML_BLOCK]: () => (<Html {...attributes}>{children}</Html>),
    [CODE_BLOCK]: () => (<Code {...attributes}>{children}</Code>),
    [BLOCK_QUOTE]: () => (<Quote {...attributes}>{children}</Quote>),
    [OL_LIST]: () => (<OLList {...attributes}>{children}</OLList>),
    [UL_LIST]: () => (<ULList {...attributes}>{children}</ULList>),
    [LIST_ITEM]: () => (<ListItem {...attributes}>{children}</ListItem>),
    [IMAGE]: () => (<ImageElement {...props} />),
    [HR]: () => (<HorizontalRule {...attributes}>{children}</HorizontalRule>),
    [HTML_INLINE]: () => (<Inline {...attributes} content={data.content}>{children}</Inline>),
    default: () => {
      console.log(`Didn't know how to render ${JSON.stringify(element, null, 2)}`);
      return <p {...attributes}>{children}</p>;
    }
  };
  const elementRenderer = customElements
    ? { ...baseElementRenderer, ...customElements(attributes, children, element) }
    : baseElementRenderer;
  return (elementRenderer[type] || elementRenderer.default)();
};

Element.propTypes = {
  children: PropTypes.node,
  element: PropTypes.shape({
    data: PropTypes.object,
    type: PropTypes.string
  }),
  attributes: PropTypes.any
};

export default Element;
