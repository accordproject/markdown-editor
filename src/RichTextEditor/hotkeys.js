const HOTKEYS = {
  'mod+b': 'bold',
  'mod+i': 'italic',
  'mod+`': 'code'
};

export default HOTKEYS;
