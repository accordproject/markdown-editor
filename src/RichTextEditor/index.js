import React, {
  useCallback, useMemo, useState
} from 'react';
import { CiceroMarkTransformer } from '@accordproject/markdown-cicero';
import { HtmlTransformer } from '@accordproject/markdown-html';
import { SlateTransformer } from '@accordproject/markdown-slate';
import isHotkey from 'is-hotkey';
import { Editable, withReact, Slate } from 'slate-react';
import { Node, createEditor, Transforms } from 'slate';
import { withHistory } from 'slate-history';
import PropTypes from 'prop-types';
import HOTKEYS, {
  ENTER, ENTER_BREAKS, formattingHotKeys
} from './utilities/hotkeys';
import withSchema from './utilities/schema';
import Element from './components';
import Leaf from './components/Leaf';
import { toggleMark, toggleBlock } from './utilities/toolbarHelpers';
import { withImages, insertImage } from './plugins/withImages';
import { withLinks, isSelectionLinkBody } from './plugins/withLinks';
import { withHtml } from './plugins/withHtml';
import FormatBar from './FormattingToolbar';

const RichTextEditor = (props) => {
  const [showLinkModal, setShowLinkModal] = useState(false);
  const { augmentEditor } = props;
  const renderLeaf = useCallback(props => <Leaf {...props} />, []);
  const editor = useMemo(() => {
    if (augmentEditor) {
      return augmentEditor(
        withLinks(withHtml(withImages(withSchema(withHistory(withReact(createEditor()))))))
      );
    }
    return withLinks(withHtml(withImages(withSchema(withHistory(withReact(createEditor()))))));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const renderElement = useCallback((slateProps) => {
    const elementProps = { ...slateProps, customElements: props.customElements };
    return (<Element {...elementProps} />);
  }, [props.customElements]);

  const hotkeyActions = {
    mark: code => toggleMark(editor, code),
    block: code => toggleBlock(editor, code),
    image: () => {
      const url = window.prompt('Enter the URL of the image:');
      if (url) {
        insertImage(editor, url);
      }
    },
    special: (code) => {
      if (code === 'undo') return editor.undo();
      return editor.redo();
    },
    link: () => {
      setShowLinkModal(true);
    },
  };

  const { isEditable, canBeFormatted } = props;

  const onKeyDown = useCallback((event) => {
    if (isHotkey(ENTER, event)) {
      // eslint-disable-next-line no-restricted-syntax
      for (const [n] of Node.ancestors(editor, editor.selection.focus.path, { reverse: true })) {
        if (ENTER_BREAKS[n.type]) {
          const currNode = Node.get(editor, editor.selection.focus.path);
          if (currNode.object === 'text' && currNode.text === '') {
            Transforms.unwrapNodes(editor, { match: n => ENTER_BREAKS[n.type], split: true });
            return;
          }
        }
      }
    }

    const canFormat = canBeFormatted(editor);
    const isFormatEvent = () => formattingHotKeys.some(hotkey => isHotkey(hotkey, event));
    if (!canFormat && isFormatEvent()) {
      event.preventDefault();
      return;
    }

    const hotkeys = Object.keys(HOTKEYS);
    hotkeys.forEach((hotkey) => {
      if (isHotkey(hotkey, event)) {
        event.preventDefault();
        const { code, type } = HOTKEYS[hotkey];
        hotkeyActions[type](code);
      }
    });
  }, [canBeFormatted, editor, hotkeyActions]);

  const onBeforeInput = useCallback((event) => {
    const canEdit = isEditable(editor, event);
    if (!canEdit) {
      event.preventDefault();
    }
  }, [editor, isEditable]);

  const handleCopyOrCut = useCallback((event) => {
    event.preventDefault();
    const slateTransformer = new SlateTransformer();
    const htmlTransformer = new HtmlTransformer();
    const ciceroMarkTransformer = new CiceroMarkTransformer();

    // The "JSON" from Slate is immutable
    // https://github.com/ianstormtaylor/slate/issues/3577
    // We need to take a functional approach
    // https://github.com/accordproject/markdown-transform/issues/203
    const SLATE_CHILDREN = JSON.parse(JSON.stringify(Node.fragment(editor, editor.selection)));
    const SLATE_DOM = {
      object: 'value',
      document: {
        object: 'document',
        data: {},
        children: SLATE_CHILDREN
      }
    };

    const CICERO_MARK_DOM = slateTransformer.toCiceroMark(SLATE_DOM);
    const HTML_DOM = htmlTransformer.toHtml(CICERO_MARK_DOM);
    const MARKDOWN_TEXT = ciceroMarkTransformer.toMarkdown(CICERO_MARK_DOM);

    event.clipboardData.setData('text/html', HTML_DOM);
    event.clipboardData.setData('text/plain', MARKDOWN_TEXT);
  }, [editor]);

  const onChange = (value) => {
    if (props.readOnly) return;
    props.onChange(value, editor);
    const { selection } = editor;
    if (selection && isSelectionLinkBody(editor)) {
      setShowLinkModal(true);
    }
  };

  return (
    <Slate editor={editor} value={props.value} onChange={onChange}>
      { !props.readOnly
        && <FormatBar
        lockText={props.lockText}
        canBeFormatted={props.canBeFormatted}
        showLinkModal={showLinkModal}
        setShowLinkModal={setShowLinkModal}
        /> }
      <Editable
        className="ap-rich-text-editor"
        readOnly={props.readOnly}
        renderElement={renderElement}
        renderLeaf={renderLeaf}
        placeholder="Enter some rich textâ€¦"
        spellCheck
        autoFocus
        onKeyDown={onKeyDown}
        onDOMBeforeInput={onBeforeInput}
        onCopy={event => handleCopyOrCut(event)}
        onCut={event => handleCopyOrCut(event)}
      />
    </Slate>
  );
};

/**
 * The property types for this component
 */
RichTextEditor.propTypes = {
  /* Initial contents for the editor (markdown text) */
  value: PropTypes.array.isRequired,
  /* Props for the editor */
  editorProps: PropTypes.object.isRequired,
  /* A callback that receives the markdown text */
  onChange: PropTypes.func.isRequired,
  /* Boolean to make editor read-only (uneditable) or not (editable) */
  readOnly: PropTypes.bool,
  /* Boolean to lock non variable text */
  lockText: PropTypes.bool,
  /* Higher order function to augment the editor methods */
  augmentEditor: PropTypes.func,
  /* Array of plugins passed in for the editor */
  customElements: PropTypes.object,
  /* A method that determines if current edit should be allowed */
  isEditable: PropTypes.func,
  /* A method that determines if current formatting change should be allowed */
  canBeFormatted: PropTypes.func,
};

RichTextEditor.defaultProps = {
  isEditable: () => true,
  canBeFormatted: () => true,
};


export default RichTextEditor;
