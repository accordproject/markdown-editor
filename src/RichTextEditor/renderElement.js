import React from 'react';
import Heading from './components/heading';
import ImageElement from './components/image';

import * as SCHEMA from './schema';

const Element = (props) => {
  const { attributes, children, element } = props;

  switch (element.type) {
    case SCHEMA.PARAGRAPH:
      return <p {...attributes}>{children}</p>;
    case SCHEMA.H1:
      return <Heading as="h1" {...attributes}>{children}</Heading>;
    case SCHEMA.H2:
      return <Heading as="h2" {...attributes}>{children}</Heading>;
    case SCHEMA.H3:
      return <Heading as="h3" {...attributes}>{children}</Heading>;
    case 'heading_four':
      return <Heading as="h4" {...attributes}>{children}</Heading>;
    case 'heading_five':
      return <Heading as="h5" {...attributes}>{children}</Heading>;
    case 'heading_six':
      return <Heading as="h6" {...attributes}>{children}</Heading>;
    case 'horizontal_rule':
      return <div className="hr" {...attributes}>{children}</div>;
    case 'code_block':
      return <pre {...attributes}>{children}</pre>;
    case 'html_block':
      return <pre className="html_block" {...attributes}>{children}</pre>;
    case 'block_quote':
      return <blockquote {...attributes}>{children}</blockquote>;
    case 'ul_list':
      return <ul {...attributes}>{children}</ul>;
    case 'heading_one':
      return <h1 {...attributes}>{children}</h1>;
    case 'heading_two':
      return <h2 {...attributes}>{children}</h2>;
    case 'list_item':
      return <li {...attributes}>{children}</li>;
    case 'ol_list':
      return <ol {...attributes}>{children}</ol>;
    case 'link':
      return <a {...attributes} href={element.data.href}>{children}</a>;
    case 'image':
      return <ImageElement {...props} />;
    // case 'image':
    //   return <img {...attributes} alt={element.data.title} src={element.data.href}>{children}</img>;
    case 'html_inline':
      return <span className='html_inline' {...attributes}>{element.data.content}{children}</span>;
    case 'softbreak':
      return <span className='softbreak' {...attributes}> {children}</span>;
    case 'linebreak':
      return <br className='linebreak' {...attributes}>{children}</br>;
    default:
      console.log(`Didn't know how to render ${JSON.stringify(element, null, 2)}`);
      return <p {...attributes}>{children}</p>;
  }
};

export default Element;
