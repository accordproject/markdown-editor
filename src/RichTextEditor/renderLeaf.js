import React from 'react';

const Leaf = ({ attributes, children, leaf }) => {
  if (leaf.bold) {
    children = <strong>{children}</strong>;
  }

  if (leaf.code) {
    children = <code>{children}</code>;
  }

  if (leaf.italic) {
    children = <em>{children}</em>;
  }

  return <span {...attributes}>{children}</span>;
};

export default Leaf;
