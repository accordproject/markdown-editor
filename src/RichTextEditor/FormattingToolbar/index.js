import React, { useRef } from 'react';
import { InsertImageButton } from '../components/withImages';
import ToolbarMenu from './ToolbarMenu';
import FormatButton from './FormatButton';
import HistoryButton from './HistoryButton';
import HyperlinkButton from './HyperlinkButton';
import StyleDropdown from './StyleDropdown';
import { insertLink, isLinkActive } from '../components/withLinks';
import {
  toggleBlock, isBlockActive,
  toggleMark, isMarkActive,
  toggleHistory
} from '../utilities/toolbarHelpers';
import {
  bold, italic, code,
  quote, olist, ulist,
  image, link, undo, redo,
  Separator
} from '../components/icons';

const mark = { toggleFunc: toggleMark, activeFunc: isMarkActive };
const block = { toggleFunc: toggleBlock, activeFunc: isBlockActive };
const history = { toggleFunc: toggleHistory };
const hyperlink = { toggleFunc: insertLink, activeFunc: isLinkActive };

const FormattingToolbar = () => {
  const linkButtonRef = useRef();
  return (
    <ToolbarMenu>
      <StyleDropdown />
      <Separator />
      <FormatButton {...mark} {...bold} />
      <FormatButton {...mark} {...italic} />
      <FormatButton {...mark} {...code} />
      <Separator />
      <FormatButton {...block} {...quote} />
      <FormatButton {...block} {...olist} />
      <FormatButton {...block} {...ulist} />
      <Separator />
      <HistoryButton {...history} {...undo} />
      <HistoryButton {...history} {...redo} />
      <Separator />
      <HyperlinkButton ref={linkButtonRef} {...hyperlink} {...link}/>
      <InsertImageButton {...image}/>
    </ToolbarMenu>
  );
};

export default FormattingToolbar;
