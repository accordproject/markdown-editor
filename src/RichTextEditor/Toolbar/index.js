import React from 'react';
import PropTypes from 'prop-types';
import { cx, css } from 'emotion';
import {
  useSlate
} from 'slate-react';
import Button from '../components/Button';
import Icon from '../components/Icon';
import Menu from '../components/Menu';
import { InsertImageButton } from '../components/Image';


import {
  toggleMark, toggleBlock, isBlockActive, isMarkActive
} from '../utilities/toolbarHelpers';

// eslint-disable-next-line react/display-name
const Toolbar = React.forwardRef(({ className, ...props }, ref) => (
    <Menu
      {...props}
      ref={ref}
      className={cx(
        className,
        css`
          position: relative;
          padding: 1px 18px 17px;
          margin: 0 -20px;
          border-bottom: 2px solid #eee;
          margin-bottom: 20px;
        `
      )}
    />
));

Toolbar.propTypes = {
  className: PropTypes.string,
};

const BlockButton = ({ format, icon }) => {
  const editor = useSlate();
  return (
      <Button
        active={isBlockActive(editor, format)}
        onMouseDown={(event) => {
          event.preventDefault();
          toggleBlock(editor, format);
        }}
      >
        <Icon>{icon}</Icon>
      </Button>
  );
};

BlockButton.propTypes = {
  format: PropTypes.string,
  icon: PropTypes.string
};

const MarkButton = ({ format, icon }) => {
  const editor = useSlate();
  return (
      <Button
        active={isMarkActive(editor, format)}
        onMouseDown={(event) => {
          event.preventDefault();
          toggleMark(editor, format);
        }}
      >
        <Icon>{icon}</Icon>
      </Button>
  );
};

MarkButton.propTypes = {
  format: PropTypes.string,
  icon: PropTypes.string
};

const FormatBar = () => (
  <Toolbar>
    <MarkButton format="bold" icon="format_bold" />
    <MarkButton format="italic" icon="format_italic" />
    <MarkButton format="code" icon="code" />
    <BlockButton format="heading_one" icon="looks_one" />
    <BlockButton format="heading_two" icon="looks_two" />
    <BlockButton format="block_quote" icon="format_quote" />
    <BlockButton format="ol_list" icon="format_list_numbered" />
    <BlockButton format="ul_list" icon="format_list_bulleted" />
    <InsertImageButton />
  </Toolbar>
);

export default FormatBar;
