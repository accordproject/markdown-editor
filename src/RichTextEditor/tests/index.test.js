import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { SlateTransformer } from '@accordproject/markdown-slate';
import RichTextEditor from '../index';

const slateTransformer = new SlateTransformer();
Enzyme.configure({ adapter: new Adapter() });
const defaultMarkdown = `# My Heading

This is text. This is *italic* text. This is **bold** text. This is a [link](https://clause.io). This is \`inline code\`.

This is ***bold and italic*** text.

## Breaks
This is a  
hard break.

This is a
softbreak.

---

This ^^^^ is a thematic break

![ap_logo](https://docs.accordproject.org/docs/assets/020/template.png "AP triangle")

> This is a quote.
## Heading Two
This is more text.

Below is a code block:

\`\`\` javascript
this is my great
code
\`\`\`

Ordered lists:

1. one
1. two
1. three

Or:

* apples
* pears
* peaches

### Sub heading

This is more text.

<custom>
This is an html block.
</custom>

And this is <variable>an HTML inline</variable>.

# H1
## H2
### H3
#### H4
#### H5
##### H6

Fin.
`;
const mockOnChange = jest.fn();

describe('<RichTextEditor />', () => {
  describe('on initialization', () => {
    it('renders page correctly', () => {
      const slate = slateTransformer.fromMarkdown(defaultMarkdown);

      const component = shallow(
        <RichTextEditor
            className="ap-rich-text-editor"
            readOnly={false}
            lockText={true}
            value={slate.document.children}
            onChange={mockOnChange}
        />
      );
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
  });
});
