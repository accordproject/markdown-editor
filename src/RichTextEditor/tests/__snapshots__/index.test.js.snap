// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichTextEditor /> on initialization renders page correctly 1`] = `
<Slate
  editor={
    Object {
      "addMark": [Function],
      "apply": [Function],
      "children": Array [],
      "deleteBackward": [Function],
      "deleteForward": [Function],
      "deleteFragment": [Function],
      "history": Object {
        "redos": Array [],
        "undos": Array [],
      },
      "insertBreak": [Function],
      "insertData": [Function],
      "insertFragment": [Function],
      "insertNode": [Function],
      "insertText": [Function],
      "isInline": [Function],
      "isVoid": [Function],
      "marks": null,
      "normalizeNode": [Function],
      "onChange": [Function],
      "operations": Array [],
      "redo": [Function],
      "removeMark": [Function],
      "selection": null,
      "undo": [Function],
    }
  }
  onChange={[Function]}
  value={
    Array [
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "My Heading",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_one",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This is text. This is ",
          },
          Object {
            "italic": true,
            "object": "text",
            "text": "italic",
          },
          Object {
            "object": "text",
            "text": " text. This is ",
          },
          Object {
            "bold": true,
            "object": "text",
            "text": "bold",
          },
          Object {
            "object": "text",
            "text": " text. This is a ",
          },
          Object {
            "children": Array [
              Object {
                "object": "text",
                "text": "link",
              },
            ],
            "data": Object {
              "href": "https://clause.io",
              "title": "",
            },
            "object": "inline",
            "type": "link",
          },
          Object {
            "object": "text",
            "text": ". This is ",
          },
          Object {
            "code": true,
            "object": "text",
            "text": "inline code",
          },
          Object {
            "object": "text",
            "text": ".",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This is ",
          },
          Object {
            "bold": true,
            "italic": true,
            "object": "text",
            "text": "bold and italic",
          },
          Object {
            "object": "text",
            "text": " text.",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Breaks",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_two",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This is a",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "data": Object {},
            "object": "inline",
            "type": "linebreak",
          },
          Object {
            "object": "text",
            "text": "hard break.",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This is a",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "data": Object {},
            "object": "inline",
            "type": "softbreak",
          },
          Object {
            "object": "text",
            "text": "softbreak.",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "text": "",
          },
        ],
        "data": Object {},
        "hr": true,
        "object": "block",
        "type": "horizontal_rule",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This ^^^^ is a thematic break",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "object": "text",
                "text": "",
              },
            ],
            "data": Object {
              "href": "https://docs.accordproject.org/docs/assets/020/template.png",
              "title": "AP triangle",
            },
            "object": "inline",
            "type": "image",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "block_quote": true,
        "children": Array [
          Object {
            "children": Array [
              Object {
                "object": "text",
                "text": "This is a quote.",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "paragraph",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "block_quote",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Heading Two",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_two",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This is more text.",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Below is a code block:",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "object": "text",
                "text": "this is my great
code
",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "paragraph",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "code_block",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Ordered lists:",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "object": "text",
                    "text": "one",
                  },
                ],
                "data": Object {},
                "object": "block",
                "type": "paragraph",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "list_item",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "object": "text",
                    "text": "two",
                  },
                ],
                "data": Object {},
                "object": "block",
                "type": "paragraph",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "list_item",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "object": "text",
                    "text": "three",
                  },
                ],
                "data": Object {},
                "object": "block",
                "type": "paragraph",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "list_item",
          },
        ],
        "data": Object {
          "delimiter": "period",
          "start": "1",
          "tight": "true",
        },
        "object": "block",
        "type": "ol_list",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Or:",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "object": "text",
                    "text": "apples",
                  },
                ],
                "data": Object {},
                "object": "block",
                "type": "paragraph",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "list_item",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "object": "text",
                    "text": "pears",
                  },
                ],
                "data": Object {},
                "object": "block",
                "type": "paragraph",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "list_item",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "object": "text",
                    "text": "peaches",
                  },
                ],
                "data": Object {},
                "object": "block",
                "type": "paragraph",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "list_item",
          },
        ],
        "data": Object {
          "delimiter": undefined,
          "start": undefined,
          "tight": "true",
        },
        "object": "block",
        "type": "ul_list",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Sub heading",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_three",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "This is more text.",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "html": true,
                "object": "text",
                "text": "<custom>
This is an html block.
</custom>",
              },
            ],
            "data": Object {},
            "object": "block",
            "type": "paragraph",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "html_block",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "And this is ",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "data": Object {
              "content": "<variable>",
            },
            "object": "inline",
            "type": "html_inline",
          },
          Object {
            "object": "text",
            "text": "an HTML inline",
          },
          Object {
            "children": Array [
              Object {
                "text": "",
              },
            ],
            "data": Object {
              "content": "</variable>",
            },
            "object": "inline",
            "type": "html_inline",
          },
          Object {
            "object": "text",
            "text": ".",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "H1",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_one",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "H2",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_two",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "H3",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_three",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "H4",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_four",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "H5",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_four",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "H6",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "heading_five",
      },
      Object {
        "children": Array [
          Object {
            "object": "text",
            "text": "Fin.",
          },
        ],
        "data": Object {},
        "object": "block",
        "type": "paragraph",
      },
    ]
  }
>
  <FormattingToolbar
    canBeFormatted={[Function]}
    lockText={true}
  />
  <Editable
    autoFocus={true}
    onCopy={[Function]}
    onCut={[Function]}
    onKeyDown={[Function]}
    placeholder="Enter some rich text…"
    readOnly={false}
    renderElement={[Function]}
    renderLeaf={[Function]}
    spellCheck={true}
  />
</Slate>
`;
